### PR TITLE
Add support for Gradle Configuration Cache

### LINKED ISSUE #626: java.util.ConcurrentModificationException (no error message)
I think there is a bug with the realm plugin, when task: :compileKotlinIosSimulatorArm64 is running, i am getting the following:

```
* What went wrong:
Execution failed for task ':compileKotlinIosSimulatorArm64'.
> java.util.ConcurrentModificationException (no error message)
```

Sometimes it works and sometimes it doesn't, i guess due to the concurrency.

```
i: [Realm Compiler Plugin] 2022-01-07T09:43:13.830916Z Processing class Alias

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.831753Z String property named stringValue is nullable false

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836332Z Processing class Category

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836418Z Object property named wallpaper is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836575Z String property named title is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836598Z String property named trackingName is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836614Z String property named path is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836630Z Boolean property named isHidden is nullable false

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836649Z RealmList property named cells is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836971Z RealmList property named showcases is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.836995Z RealmList property named sections is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837015Z RealmList property named aliases is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837032Z Int property named rawCategoryType is nullable false

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837587Z Processing class Cell

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837606Z String property named id is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837627Z Int property named rawType is nullable false

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837910Z Processing class Section

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837930Z Int property named size is nullable false

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837959Z Boolean property named showTitle is nullable false

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837978Z String property named title is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.837991Z String property named viewAllUrl is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838002Z String property named trackingName is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838029Z RealmList property named cells is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838349Z Processing class Wallpaper

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838378Z String property named rawHighlightColour is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838399Z String property named rawBackgroundColour is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838419Z Int property named rawTheme is nullable false

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838443Z String property named topImage is nullable true

i: [Realm Compiler Plugin] 2022-01-07T09:43:13.838455Z String property named bottomImage is nullable true
```

Here is the same with --stacktrace
```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileKotlinIos'.
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:94)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:402)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:389)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:382)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:368)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)
Caused by: java.util.ConcurrentModificationException
	at org.gradle.api.internal.DefaultDomainObjectCollection$IteratorImpl.next(DefaultDomainObjectCollection.java:475)
	at io.realm.gradle.RealmAnalytics.jsonPayload(RealmAnalytics.kt:112)
	at io.realm.gradle.RealmAnalytics.sendMetricIfNeeded(RealmAnalytics.kt:90)
	at io.realm.gradle.RealmAnalytics.afterExecute(RealmAnalytics.kt:82)
	at jdk.internal.reflect.GeneratedMethodAccessor184.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:464)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:446)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:61)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:434)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:421)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:43)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:245)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:157)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy89.afterExecute(Unknown Source)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:91)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:402)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:389)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:382)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:368)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)
```

### PR DESCRIPTION
**Pull request by cmelchior**
Closes #626 

This should speed up builds with configuration cache enabled and should also be a work-around for https://github.com/realm/realm-kotlin/issues/626

TODO:
- [x] Our own repo currently report a number of errors for the Configuration Cache. Timebox to see if this can be easily fixed (= They are not, address this in another PR)
- [x] Attempt to verify that race condition is gone from #626. I haven't been able to reproduce the race in the first place.
- [x] CHANGELOG

### CODE REVIEW

**Comment by cmelchior** [comment]
This PR is now ready for review. I tried to reproduce the crash in #626, but couldn't. At least not when building our samples with parallel builds enabled. However, judging from the code around the crash, I _think_ this PR should also fix that issue.

I also tried to enable the configuration cache for our own build, but that spiraled into massive changes with no end in sight, so I'm leaving that as a future improvement.

**Review by rorbech — APPROVED** [review]
I guess it is a major improvement that we support configuration cache, but it is really hard to evaluate if this is the right approach. 

Seems like what we really want for collecting metrics is actually a [shared build service](https://docs.gradle.org/7.4/userguide/build_services.html)(available from Gradle 7.4) that can collect the stats for each task and submit them when closing.

File: packages/gradle-plugin/src/main/kotlin/io/realm/gradle/analytics/RealmAnalytics.kt [hunk 1]
```diff
@@ -164,20 +175,21 @@ internal class RealmAnalytics : TaskExecutionAdapter() {
     @Suppress("TooGenericExceptionCaught")
     private fun sendAnalytics(json: String, logger: Logger) {
         try {
-            logger.log(LogLevel.DEBUG, "REALM ANALYTICS: sending payload\n$json")
+            logger.debug("Sending analytics payload\n$json")
+            // TODO Consider turning this into a daemon thread so we don't block the build for
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/gradle-plugin/src/main/kotlin/io/realm/gradle/analytics/RealmAnalytics.kt:179]: Yup, but didn't wanted to fill to much in this PR. Will create an issue for tracking it.

File: packages/gradle-plugin/src/main/kotlin/io/realm/gradle/analytics/RealmAnalytics.kt [hunk 2]
```diff
@@ -164,20 +175,21 @@ internal class RealmAnalytics : TaskExecutionAdapter() {
...
             } catch (e: InterruptedException) {
-                logger.log(LogLevel.DEBUG, "REALM ANALYTICS: transfer interrupted")
+                logger.debug("Sending analytics was interrupted.")
                 pool.shutdownNow()
             }
         } catch (e: Exception) {
             // Analytics failing for any reason should not crash the build
-            logger.log(LogLevel.DEBUG, "REALM ANALYTICS: transfer failed")
-            System.err.println("Could not send analytics: $e")
+            logger.warn("Error when sending: $e")
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/gradle-plugin/src/main/kotlin/io/realm/gradle/analytics/RealmAnalytics.kt:192]: True, this was mostly for debugging, but `warn` is probably too strong. I'll move it back down to `debug`

**Pull request merged**